### DISCUSSION TITLE
10/11/2022 Updates

### DISCUSSION THREAD

**Discussion by wiverson** [wiverson/maven-jpackage-template#84] [Announcements]
Notable Recent Updates

- Switched to the excellent [AtlantaFX](https://github.com/mkpaz/atlantafx) theme
- Redid the way installers build (single script, GitHub Action matrix) and version (timestamp -> run id)
- Now uses JDKs w/JavaFX built-in (e.g. Liberica or Zulu), vastly simplifying upgrades & config
- Fix for compile scope bug that was adding test dependencies to build ([thanks ddark008](https://github.com/wiverson/maven-jpackage-template/commit/7253a816cbfdd47a7b68e046daf23611b3f57dd3))!
- Fix for Windows auto-start script ([thanks ddark008](https://github.com/wiverson/maven-jpackage-template/pull/68/commits/aaf99858481fd145a15485066603a7682846de68))!

At bit more info on the installer build rework:

The previous version used a separate GitHub Action to create each installer, and then tried to mangle a timestamp into the version number and file name. This caused several issues, including:

- Trying to mash the timestamp-based version number into a version acceptable to Windows installer version just didn't work. This would occasionally result in invalid Windows version numbers, which is unacceptable.
- The resulting timestamp filenames were not user friendly. Names like "My App 1.0.9827937483492.dmg" are much harder for a user to parse than "My App 1.0.23"
- The various apps didn't quite match versions across platforms - ugh.

So, I wound up merging the installer scripts into a single job. The different OSes are now handled via a GitHub Actions matrix. The job number in turn generates the app version number. This offers many advantages:

- App versions now match across all generated platforms
- App versions are now simple numbers (e.g. My App 1.0.23)
- If a sub-build fails, it will automatically cancel the other builds

In addition, the build script now looks to see if the four GitHub Secrets to sign, notarize and staple macOS builds are present. If the Secrets are present, it will perform the necessary steps, but if they are not it will just skip the build. As an aside, this meant I had to sort out the syntax for combining conditionals in GitHub Action YAML which might be interesting for a lot of other uses.

At this point, we are still waiting for [GitHub Actions to support Apple Silicon](https://github.com/actions/runner-images/issues/2187) runners. You can still run mvn install to generate an M1-based installer, and if you use a [macOS self-hosted runner](https://www.pantsbuild.org/docs/ci-for-macos-on-arm64) you should be able to get things to work.

Thanks to everyone for their support!

**Comment by PerryCameron** [comment]
Nice work! Thanks for this project Will!

**Comment by lubriBoy** [comment]
Still improving, many thanks for the project Will !!!
